gnuradio-runtime/python/gnuradio/gr/bindings/block_gateway_python.cc

```diff
@@ -7,17 +7,200 @@
+                                              uint64_t abs_end,
+                                              const pmt::pmt_t& key)
+    {
+        std::vector<gr::tag_t> tags;
+        gr::block::get_tags_in_range(tags, which_input, abs_start, abs_end, key);
+        return tags;
+    }
+
+    std::vector<gr::tag_t>
+    _get_tags_in_window(unsigned int which_input, uint64_t rel_start, uint64_t rel_end)
+    {
+        std::vector<gr::tag_t> tags;
+        gr::block::get_tags_in_window(tags, which_input, rel_start, rel_end);
+        return tags;
+    }
+
+    std::vector<gr::tag_t> _get_tags_in_window(unsigned int which_input,
+                                               uint64_t rel_start,
+                                               uint64_t rel_end,
+                                               const pmt::pmt_t& key)
+    {
+        std::vector<gr::tag_t> tags;
+        gr::block::get_tags_in_window(tags, which_input, rel_start, rel_end, key);
+        return tags;
+    }
+
+    /*******************************************************************
+     * Overloads for various scheduler-called functions
+     ******************************************************************/
+    void forecast(int noutput_items, gr_vector_int& ninput_items_required) override
+    {
+        py::gil_scoped_acquire acquire;
+
+        py::object ret_ninput_items_required = _py_handle.attr("handle_forecast")(
+            noutput_items, ninput_items_required.size());
+        ninput_items_required = ret_ninput_items_required.cast<std::vector<int>>();
+    }
+
+    int general_work(int noutput_items,
+                     gr_vector_int& ninput_items,
+                     gr_vector_const_void_star& input_items,
+                     gr_vector_void_star& output_items) override
+    {
+        py::gil_scoped_acquire acquire;
+
+        py::object ret = _py_handle.attr("handle_general_work")(
+            noutput_items, ninput_items, input_items, output_items);
+
+        return ret.cast<int>();
+        ;
+    }
+
+    bool start(void) override
+    {
+        py::gil_scoped_acquire acquire;
+
+        py::object ret = _py_handle.attr("start")();
+        return ret.cast<bool>();
+    }
+
+    bool stop(void) override
+    {
+        py::gil_scoped_acquire acquire;
+
+        py::object ret = _py_handle.attr("stop")();
+        return ret.cast<bool>();
+    }
+
+    void set_msg_handler_pybind(pmt::pmt_t which_port, std::string& handler_name)
+    {
+        if (msg_queue.find(which_port) == msg_queue.end()) {
+            throw std::runtime_error(
+                "attempt to set_msg_handler_pybind() on invalid input message port!");
+        }
+        d_msg_handlers_pybind[which_port] = handler_name;
+    }
+
+protected:
+    // Message handlers back into python using pybind API
+    typedef std::map<pmt::pmt_t, std::string, pmt::comparator> msg_handlers_pybind_t;
+    msg_handlers_pybind_t d_msg_handlers_pybind;
```

Comment:
why is this protected? Can this class be sub-classed?

FWIW, modern C++ prefers `using`:
```suggestion
    // Message handlers back into python using pybind API
    using msg_handlers_pybind_t = std::map<pmt::pmt_t, std::string, pmt::comparator>;
    msg_handlers_pybind_t d_msg_handlers_pybind;
```